src/event-cache/EventStore.ts

```diff
@@ -0,0 +1,56 @@
+import { RumEvent } from '../dispatch/dataplane';
+
+// store events by calling context.record(type, event, key)
+// get events by key with context.getEvent(key)
+// events are evicted from store when removed from EventCache
+export class EventStore {
+    // maps custom keys to RUM events
+    private events = new Map<string, RumEvent>();
+    // maps event ids to custom keys
+    private keys = new Map<string, string>();
+
+    get(key: string) {
+        return this.events.get(key);
+    }
+
+    getById(id: string) {
+        const key = this.keys.get(id);
+        if (key) {
+            return this.events.get(key);
+        }
+    }
+
+    put(key: string, event: RumEvent) {
+        const dup = this.get(key);
+        if (dup) {
+            this.keys.delete(dup.id);
+        }
+        this.keys.set(event.id, key);
+        this.events.set(key, event);
+    }
+
+    evict(key: string) {
```

Comment:
question: Is there a potential for a condition between the time events are evicted vs when the event is already dispatched?

Reply:
There is the edge case that when LCP attribution is reported, then resource and navigation events have already been batched and evicted. My understanding is that this is rare, since dispatch interval is 5 seconds and LCP is usually reported pretty quickly. 

However, @qhanam and I were discussing using a different eviction strategy to address this edge case and reduce the manageability burden. Storing "first N" events would likely capture all the LCP related events and reduce the overall size of this PR. "Most recent N" is similar, with a bit more overhead. 

Giving this some thought and will raise another revision.